[Optimus] Replace torch rms norm with quack rms norm kernel

mengluy0125 opened the pull request:
Summary: We observe significant perf gain to replace torch rms norm with the quack rms norm kernel (D80317390). We do this inside compiler, thus it can be reused across models with at least B200 GPUs

Test Plan:
### unit test

```
buck2 test mode/opt  -c fbcode.nvcc_arch=b200a -c fbcode.enable_gpu_sections=true -c fbcode.platform010_cuda_version=12.8 -m "ovr_config//third-party/pypi/nvidia-cutlass-dsl/constraints:4.1.0" //caffe2/test/inductor:kernel_optimization -- test_replace_rms_norm_with_quack
```

Buck UI: https://www.internalfb.com/buck2/c29f2d25-b674-44c5-b504-a475b209384c
Test UI: https://www.internalfb.com/intern/testinfra/testrun/12666374058307987
Network: Up: 38KiB  Down: 12MiB  (reSessionID-0cbaccbb-7f2c-4651-9fc2-c13760f0e8f7)
Analyzing targets. Remaining     0/287
Executing actions. Remaining     0/21254                                                                      11.9s exec time total
Command: test.     Finished 5 local, 1 cache (17% hit)                                                        7.2s exec time cached (60%)
Time elapsed: 24.5s
Tests finished: Pass 1. Fail 0. Fatal 0. Skip 0. Build failure 0

### EMS

Enable the following flag
```
        torch._inductor.config.pre_grad_fusion_options = {
            "use_custom_rmsnorm_kernel_pass": {"quack": True}
        }
```

```
buck2 run mode/opt -c fbcode.platform010_cuda_version=12.8 aps_models/ads/ecosystem/tooling/tools/efficient_module_suite/benchmark:omnifm_perf_benchmark -- benchmark-with-prod-model --prod_config mast_omnifm_v3_B200_1kgpu --prod_config_override prod_config_override_jointarch --batch_size 1152 --enable_pt2 True --mfu_profile_module uniarch_layers.0.seq_interaction_layer --enable_ac True
```


without quack kernel

```
| Metric                | Value        |
|:----------------------|:-------------|
| Batch size            | 1152         |
| GPU type              | B200         |
| Latency               | 74.57 ms     |
| Model size            | 1.44 GB      |
| Flops                 | 14431.98 G   |
| Flops/example         | 12.53 G      |
| TFLOPS/sec            | 193.53       |
| MFU                   | 16.54%       |
| Activation/example    | 12.27 MB     |
| CPU time total        | 224.09 ms    |
| GPU time total        | 262.45 ms    |
| Estimated avg BW      | 2534.11 GB/s |
| Estimated avg BW util | 40.87%       |
Trace link: https://our.intern.facebook.com/intern/perfdoctor/trace_view?filepath=tree/traces/efficient_module_suite/uniarch.uniarch_layers.0.seq_interaction_layer.Aug_07_22_26_33_trace.json.gz&bucket=pyper_traces
Perfetto link: https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/pyper_traces/tree/traces/efficient_module_suite/uniarch.uniarch_layers.0.seq_interaction_layer.Aug_07_22_26_33_trace.json.gz
Snapshot link: https://www.internalfb.com/pytorch_memory_visualizer/ai_efficiency/tree/gpu_snapshot/uniarch.uniarch_layers.0.seq_interaction_layer.Aug_07_22_26_33.snapshot.pickle
```

with quack kernel

```
| Metric             | Value      |
|:-------------------|:-----------|
| Batch size         | 1152       |
| GPU type           | B200       |
| Latency            | 71.72 ms   |
| Model size         | 1.44 GB    |
| Flops              | 14431.98 G |
| Flops/example      | 12.53 G    |
| TFLOPS/sec         | 201.22     |
| MFU                | 17.20%     |
| Activation/example | 12.28 MB   |
| CPU time total     | 197.22 ms  |
| GPU time total     | 237.44 ms  |
Trace link: https://our.intern.facebook.com/intern/perfdoctor/trace_view?filepath=tree/traces/efficient_module_suite/uniarch_layers.0.seq_interaction_layer.Aug_26_17_07_41_trace.json.gz&bucket=pyper_traces
Perfetto link: https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/pyper_traces/tree/traces/efficient_module_suite/uniarch_layers.0.seq_interaction_layer.Aug_26_17_07_41_trace.json.gz
Snapshot link: https://www.internalfb.com/pytorch_memory_visualizer/ai_efficiency/tree/gpu_snapshot/uniarch_layers.0.seq_interaction_layer.Aug_26_17_07_41.snapshot.pickle
```

Rollback Plan:

Differential Revision: D80492377




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @mlazos

Skylion007 commented on torch/_inductor/fx_passes/kernel_optimization.py:
```diff
@@ -0,0 +1,74 @@
+
+def print_norm_pattern(match: Match, inputs: list[Optional[torch.fx.Node]]):
+    node = match.nodes[-1]
+    log.debug(
```
This is still eagerly does the string join, either wrap it in a LazyStr or put it in a proper debugging level if

mengluy0125 committed:
[Optimus] Replace torch rms norm with quack rms norm kernel (#161630)

Summary:

We observe significant perf gain to replace torch rms norm with the quack rms norm kernel (D80317390). We do this inside compiler, thus it can be reused across models with at least B200 GPUs

Test Plan:
### unit test

```
buck2 test mode/opt  -c fbcode.nvcc_arch=b200a -c fbcode.enable_gpu_sections=true -c fbcode.platform010_cuda_version=12.8 -m "ovr_config//third-party/pypi/nvidia-cutlass-dsl/constraints:4.1.0" //caffe2/test/inductor:kernel_optimization -- test_replace_rms_norm_with_quack
```

Buck UI: https://www.internalfb.com/buck2/c29f2d25-b674-44c5-b504-a475b209384c
Test UI: https://www.internalfb.com/intern/testinfra/testrun/12666374058307987
Network: Up: 38KiB  Down: 12MiB  (reSessionID-0cbaccbb-7f2c-4651-9fc2-c13760f0e8f7)
Analyzing targets. Remaining     0/287
Executing actions. Remaining     0/21254                                                                      11.9s exec time total
Command: test.     Finished 5 local, 1 cache (17% hit)                                                        7.2s exec time cached (60%)
Time elapsed: 24.5s
Tests finished: Pass 1. Fail 0. Fatal 0. Skip 0. Build failure 0

### EMS

Enable the following flag
```
        torch._inductor.config.pre_grad_fusion_options = {
            "use_custom_rmsnorm_kernel_pass": {"quack": True}
        }
```

```
buck2 run mode/opt -c fbcode.platform010_cuda_version=12.8 aps_models/ads/ecosystem/tooling/tools/efficient_module_suite/benchmark:omnifm_perf_benchmark -- benchmark-with-prod-model --prod_config mast_omnifm_v3_B200_1kgpu --prod_config_override prod_config_override_jointarch --batch_size 1152 --enable_pt2 True --mfu_profile_module uniarch_layers.0.seq_interaction_layer --enable_ac True
```


without quack kernel

```
| Metric                | Value        |
|:----------------------|:-------------|
| Batch size            | 1152         |
| GPU type              | B200         |
| Latency               | 74.57 ms     |
| Model size            | 1.44 GB      |
| Flops                 | 14431.98 G   |
| Flops/example         | 12.53 G      |
| TFLOPS/sec            | 193.53       |
| MFU                   | 16.54%       |
| Activation/example    | 12.27 MB     |
| CPU time total        | 224.09 ms    |
| GPU time total        | 262.45 ms    |
| Estimated avg BW      | 2534.11 GB/s |
| Estimated avg BW util | 40.87%       |
Trace link: https://our.intern.facebook.com/intern/perfdoctor/trace_view?filepath=tree/traces/efficient_module_suite/uniarch.uniarch_layers.0.seq_interaction_layer.Aug_07_22_26_33_trace.json.gz&bucket=pyper_traces
Perfetto link: https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/pyper_traces/tree/traces/efficient_module_suite/uniarch.uniarch_layers.0.seq_interaction_layer.Aug_07_22_26_33_trace.json.gz
Snapshot link: https://www.internalfb.com/pytorch_memory_visualizer/ai_efficiency/tree/gpu_snapshot/uniarch.uniarch_layers.0.seq_interaction_layer.Aug_07_22_26_33.snapshot.pickle
```

with quack kernel

```
| Metric             | Value      |
|:-------------------|:-----------|
| Batch size         | 1152       |
| GPU type           | B200       |
| Latency            | 71.72 ms   |
| Model size         | 1.44 GB    |
| Flops              | 14431.98 G |
| Flops/example      | 12.53 G    |
| TFLOPS/sec         | 201.22     |
| MFU                | 17.20%     |
| Activation/example | 12.28 MB   |
| CPU time total     | 197.22 ms  |
| GPU time total     | 237.44 ms  |
Trace link: https://our.intern.facebook.com/intern/perfdoctor/trace_view?filepath=tree/traces/efficient_module_suite/uniarch_layers.0.seq_interaction_layer.Aug_26_17_07_41_trace.json.gz&bucket=pyper_traces
Perfetto link: https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/pyper_traces/tree/traces/efficient_module_suite/uniarch_layers.0.seq_interaction_layer.Aug_26_17_07_41_trace.json.gz
Snapshot link: https://www.internalfb.com/pytorch_memory_visualizer/ai_efficiency/tree/gpu_snapshot/uniarch_layers.0.seq_interaction_layer.Aug_26_17_07_41.snapshot.pickle
```

### E2E
cmf baseline (B200)
f784511811

cmf proposal

Rollback Plan:

Differential Revision: D80492377

c00w commented on torch/_inductor/kernel/quack/__init__.py:
```diff
@@ -0,0 +1,3 @@
+__version__ = "0.1.7"
```
Please also attach the license file from quack. (Otherwise I believe you're breaking the license).
